hipscat/margin_utils.py

```diff
@@ -0,0 +1,195 @@
+    #   pix ===>  4*pix+2    4*pix+1  ===>  4*(4*pix+2)+2   4*(4*pix+2)+1   4*(4*pix+1)+2   4*(4*pix+1)+1
+    #                                                4*(4*pix+2)    4*(4*pix)+3    4*(4*pix+1)
+    #                    4*pix                                4*(4*pix)+2   4*(4*pix)+1
+    #                                                                  4*(4*pix)
+    # etcetera...
+    #
+    # We can see that the edge indices follow a pattern. For example, after two
+    # subdivisions the south-east edge would consist of pixels:
+    #     4*(4*pix), 4*(4*pix)+1
+    #     4*(4*pix+1), 4*(4*pix+1)+1
+    # with the top coming from subdividing the southern, and bottom the eastern pixel.
+    # and so on, recursively, for more subdivisions. Similar patterns are identifiable
+    # with other edges.
+    #
+    # This can be compactly written as:
+    #
+    #   4*(4*(4*pix + i) + j) + k ....
+    #
+    # with i, j, k, ... being indices whose choice specifies which edge we get.
+    # For example iterating through, i = {0, 1}, j = {0, 1}, k = {0, 1} generates indices
+    # for the 8 pixels of the south-east edge for 3 subdivisions. Similarly, for
+    # the north-west edge the index values would loop through {2, 3}, etc.
+    #
+    # This can be expanded as:
+    #
+    #   4**dk*pix  +  4**(dk-1)*i + 4**(dk-2)*j + 4**(dk-3) * k + ...
+    #
+    # where dk is the number of subdivisions. E.g., for dk=3, this would equal:
+    #
+    #   4**3*pix + 4**2*i + 4**1*j + k
+    #
+    # When written with bit-shift operators, another interpretation becomes clearer:
+    #
+    #   pix << 6 + i << 4 + j << 2 + k
+    #
+    # or if we look at the binary representation of the resulting number:
+    #
+    #   [pix][ii][jj][kk]
+    #
+    # Where ii, jj, kk are the bits of _each_ of the i,j,k indices. That is, to get the
+    # list of subpixels on the edge, we bit-shift the base index pix by 2*dk to the left,
+    # and fill in the rest with all possible combinations of ii, jj, kk indices. For example,
+    # the northeast edge has index values {2, 3} = {0b10, 0b11}, so for (say) pix=8=0b1000, the
+    # northeast edge indices after two subdivisions are equal to:
+    #
+    #   0b1000 10 10 = 138
+    #   0b1000 10 11 = 139
+    #   0b1000 11 10 = 148
+    #   0b1000 11 11 = 143
+    #
+    # Note that for a given edge and dk, the suffix of each edge index does not depend on
+    # the value of pix (pix is bit-shifted to the right and added). This means it can be
+    # precomputed & stored for repeated reuse.
+    #
+    # ## Implementation:
+    #
+    # The implementation is based on the binary digit interpretation above. For a requested
+    # edge and dk subdivision level, we generate (and cache) the suffixes. Then, for a given
+    # pixel pix, the edge pixels are readily computed by left-shifting pix by 2*dk places,
+    # and summing (== or-ing) it with the suffix array.
+    #
+
+    try:
```

Comment:
This feels weird (but maybe it's pythonic?). It feels more natural / readable to say `if suffixes contains key: do something else: do something else`.

Reply:
yeah that way is probably more pythonic